Unified PDF plugin contents do not scroll

hortont424 opened the pull request:
#### cbf8b324d90f14ca4483446c751202a6e40b86a8
<pre>
Unified PDF plugin contents do not scroll
<a href="https://bugs.webkit.org/show_bug.cgi?id=264359">https://bugs.webkit.org/show_bug.cgi?id=264359</a>

Reviewed by Simon Fraser.

Hoist PDFPlugin ScrollableArea implementation to PDFPluginBase.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::installPDFDocument):
(WebKit::PDFPlugin::attemptToUnlockPDF):
(WebKit::PDFPlugin::isLocked const):
(WebKit::PDFPlugin::handleMouseEvent):
(WebKit::PDFPlugin::snapshot):
Add and adopt isLocked().

(WebKit::PDFPlugin::contentsSize const):
(WebKit::PDFPlugin::firstPageHeight const):
Add methods to pull the contents and first page sizes instead of depending
on recomputing them at various times.

(WebKit::PDFPlugin::geometryDidChange):
(WebKit::PDFPlugin::updatePDFHUDLocation):
Rename calculateSizes() to updatePDFHUDLocation() since that&apos;s the last thing it does.

(WebKit::PDFPlugin::didChangeScrollOffset):
(WebKit::PDFPlugin::notifyContentScaleFactorChanged):
(WebKit::PDFPlugin::notifyDisplayModeChanged):
(WebKit::PDFPlugin::convertFromScrollbarToContainingView const): Deleted.
(WebKit::PDFPlugin::convertFromContainingViewToScrollbar const): Deleted.
(WebKit::PDFPlugin::debugDescription const): Deleted.
(WebKit::PDFPlugin::scrollCornerRect const): Deleted.
(WebKit::PDFPlugin::enclosingScrollableArea const): Deleted.
(WebKit::PDFPlugin::scrollableAreaBoundingBox const): Deleted.
(WebKit::PDFPlugin::isActive const): Deleted.
(WebKit::PDFPlugin::forceUpdateScrollbarsOnMainThreadForPerformanceTesting const): Deleted.
(WebKit::PDFPlugin::scrollPosition const): Deleted.
(WebKit::PDFPlugin::minimumScrollPosition const): Deleted.
(WebKit::PDFPlugin::maximumScrollPosition const): Deleted.
(WebKit::PDFPlugin::scrollbarStyleChanged): Deleted.
(WebKit::PDFPlugin::deviceScaleFactor const): Deleted.
(WebKit::PDFPlugin::calculateSizes): Deleted.
(WebKit::PDFPlugin::willDetachRenderer): Deleted.
(WebKit::PDFPlugin::updateControlTints): Deleted.
(WebKit::PDFPlugin::setScrollOffset): Deleted.
Move ScrollableArea implementation to PDFPluginBase.

(WebKit::PDFPlugin::updateScrollbars):
(WebKit::PDFPlugin::createScrollbar):
(WebKit::PDFPlugin::destroyScrollbar):
Keep scrollbar *layer* code in PDFPlugin.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::scrollPositionForTesting const):
(WebKit::PDFPluginBase::willDetachRenderer): Deleted.
(WebKit::PDFPluginBase::updateControlTints): Deleted.

(WebKit::PDFPluginBase::wantsWheelEvents const):
Turn on wantsWheelEvents, since both subclasses do.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::updateControlTints):
(WebKit::PDFPluginBase::scrollCornerRect const):
(WebKit::PDFPluginBase::enclosingScrollableArea const):
(WebKit::PDFPluginBase::scrollableAreaBoundingBox const):
(WebKit::PDFPluginBase::setScrollOffset):
(WebKit::PDFPluginBase::isActive const):
(WebKit::PDFPluginBase::forceUpdateScrollbarsOnMainThreadForPerformanceTesting const):
(WebKit::PDFPluginBase::scrollPosition const):
(WebKit::PDFPluginBase::minimumScrollPosition const):
(WebKit::PDFPluginBase::maximumScrollPosition const):
(WebKit::PDFPluginBase::deviceScaleFactor const):
(WebKit::PDFPluginBase::scrollbarStyleChanged):
(WebKit::PDFPluginBase::convertFromScrollbarToContainingView const):
(WebKit::PDFPluginBase::convertFromContainingViewToScrollbar const):
(WebKit::PDFPluginBase::debugDescription const):
(WebKit::PDFPluginBase::willDetachRenderer):
(WebKit::PDFPluginBase::updateScrollbars):
(WebKit::PDFPluginBase::createScrollbar):
(WebKit::PDFPluginBase::destroyScrollbar):
Hoist ScrollableArea implementation to here from PDFPlugin.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::geometryDidChange):
(WebKit::UnifiedPDFPlugin::updateLayout):
Fold updateLayout/updateLayerHierarchy into updateLayout(), and also update scrollbars from there.

(WebKit::UnifiedPDFPlugin::scheduleRenderingUpdate):
Added; triggers a layer flush.

(WebKit::UnifiedPDFPlugin::isLocked const):
(WebKit::UnifiedPDFPlugin::contentsSize const):
(WebKit::UnifiedPDFPlugin::firstPageHeight const):
Implement these new getters.

(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):
Move the contents layer around (for now) when the scroll offset changes.

(WebKit::UnifiedPDFPlugin::invalidateScrollbarRect):
(WebKit::UnifiedPDFPlugin::invalidateScrollCornerRect):
For now, ignor scrollbar invalidation; we don&apos;t have scrollbar layers yet.

(WebKit::UnifiedPDFPlugin::handleWheelEvent):
Pass wheel events to ScrollableArea.

Canonical link: <a href="https://commits.webkit.org/270358@main">https://commits.webkit.org/270358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4c95a8523f5a8282c806e414692eb39cd900caa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25254 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3795 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26509 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27366 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23173 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5511 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1229 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25497 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2807 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21794 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27944 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2492 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28847 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/23041 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23084 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/26680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2448 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/734 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3800 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6053 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2889 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2783 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+float PDFPluginBase::deviceScaleFactor() const
+{
+    if (m_frame) {
+        if (auto* coreFrame = m_frame->coreLocalFrame()) {
```
should these be Ref / RefPtr instead of raw pointers?

cdumez replied:
It probably doesn't hurt but it is also not strictly required in this case since we're only calling trivial getters.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+    IntRect rect = scrollbarRect;
+    rect.move(scrollbar.location() - m_view->location());
+
+    return m_view->frame()->view()->convertFromRendererToContainingView(m_view->pluginElement().renderer(), rect);
```
here is a case where I'd call `protectedView()` though.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+
+IntRect PDFPluginBase::convertFromContainingViewToScrollbar(const Scrollbar& scrollbar, const IntRect& parentRect) const
+{
+    IntRect rect = m_view->frame()->view()->convertFromContainingViewToRenderer(m_view->pluginElement().renderer(), parentRect);
```
ditto.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+    IntPoint point = scrollbarPoint;
+    point.move(scrollbar.location() - m_view->location());
+
+    return m_view->frame()->view()->convertFromRendererToContainingView(m_view->pluginElement().renderer(), point);
```
ditto.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+
+IntPoint PDFPluginBase::convertFromContainingViewToScrollbar(const Scrollbar& scrollbar, const IntPoint& parentPoint) const
+{
+    IntPoint point = m_view->frame()->view()->convertFromContainingViewToRenderer(m_view->pluginElement().renderer(), parentPoint);
```
ditto.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+{
+    if (!m_frame || !m_frame->coreLocalFrame())
+        return;
+    if (auto* frameView = m_frame->coreLocalFrame()->view())
```
RefPtr

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+        m_verticalScrollbar->setFrameRect(scrollbarRect);
+    }
+
+    auto* frameView = m_frame ? m_frame->coreLocalFrame()->view() : nullptr;
```
RefPtr

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+
+Ref<Scrollbar> PDFPluginBase::createScrollbar(ScrollbarOrientation orientation)
+{
+    auto widget = Scrollbar::createNativeScrollbar(*this, orientation, ScrollbarWidth::Auto);
```
Ref

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+    // Is it ever possible that the code above and the code below can ever get at different Frames?
+    // Can't we settle on one Frame accessor?
+    if (auto* frame = m_view->frame()) {
+        if (auto* frameView = frame->view())
```
RefPtr

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -105,6 +104,16 @@
+void UnifiedPDFPlugin::scheduleRenderingUpdate()
+{
+    RefPtr frame = m_view->frame();
+    auto* page = frame->page();
```
CheckedPtr

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -1355,7 +1305,8 @@ static void dataProviderReleaseInfoCallback(void* info)
 Ref<Scrollbar> PDFPlugin::createScrollbar(ScrollbarOrientation orientation)
 {
-    auto widget = Scrollbar::createNativeScrollbar(*this, orientation, ScrollbarWidth::Auto);
+    auto scrollbar = PDFPluginBase::createScrollbar(orientation);
```
Ref

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -1780,18 +1584,28 @@ static void jsPDFDocFinalize(JSObjectRef object)
+{
+    if (isLocked() || !m_frame || !m_frame->page())
         return;
     m_frame->page()->updatePDFHUDLocation(*this, frameForHUD());
```
checkedPage()

hortont424 replied:
Doesn't currently exist! This is a WebFrame

cdumez replied:
Sorry, I got confused and thought I was in WebCore for a minute  :D
`m_frame->protectedPage()->updatePDFHUDLocation(*this, frameForHUD());` given that WebPage is refcounted.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+{
+    ASSERT(graphicsContext.invalidatingControlTints());
+
+    if (m_horizontalScrollbar)
```
```
if (RefPtr horizontalScrollbar = m_horizontalScrollbar)
    horizontalScrollbar->invalidate();
```

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+
+    if (m_horizontalScrollbar)
+        m_horizontalScrollbar->invalidate();
+    if (m_verticalScrollbar)
```
ditto.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -151,11 +160,43 @@ class PDFPluginBase : public ThreadSafeRefCounted<PDFPluginBase> {
     void invalidateRect(const WebCore::IntRect&);
 
+    // ScrollableArea functions.
+    WebCore::IntRect scrollCornerRect() const override;
```
Maybe these can all be `final`

hortont424 replied:
Some-but-not-all

hortont424 replied:
All-but-one and only because multiple inheritance be weird

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -197,6 +200,247 @@
+
+ScrollableArea* PDFPluginBase::enclosingScrollableArea() const
+{
+    // FIXME: Walk up the frame tree and look for a scrollable parent frame or RenderLayer.
```
Also PDFPluginBase has no business knowing about its ancestor tree, so it should call out to some other class to do this ancestor walk.

hortont424 replied:
But at the same time, we've gotten away without implementing this in PDFPlugin for many years, so maybe we can continue? I don't know what it's for, guess I'll have to check.

hortont424 replied:
Oh! Maybe this is why smooth keyboard scrolling doesn't work.

rr-codes replied:
oh ya this would definitely break it too

hortont424 committed:
Unified PDF plugin contents do not scroll
https://bugs.webkit.org/show_bug.cgi?id=264359

Reviewed by Simon Fraser.

Hoist PDFPlugin ScrollableArea implementation to PDFPluginBase.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::installPDFDocument):
(WebKit::PDFPlugin::attemptToUnlockPDF):
(WebKit::PDFPlugin::isLocked const):
(WebKit::PDFPlugin::handleMouseEvent):
(WebKit::PDFPlugin::snapshot):
Add and adopt isLocked().

(WebKit::PDFPlugin::contentsSize const):
(WebKit::PDFPlugin::firstPageHeight const):
Add methods to pull the contents and first page sizes instead of depending
on recomputing them at various times.

(WebKit::PDFPlugin::geometryDidChange):
(WebKit::PDFPlugin::updatePDFHUDLocation):
Rename calculateSizes() to updatePDFHUDLocation() since that's the last thing it does.

(WebKit::PDFPlugin::didChangeScrollOffset):
(WebKit::PDFPlugin::notifyContentScaleFactorChanged):
(WebKit::PDFPlugin::notifyDisplayModeChanged):
(WebKit::PDFPlugin::convertFromScrollbarToContainingView const): Deleted.
(WebKit::PDFPlugin::convertFromContainingViewToScrollbar const): Deleted.
(WebKit::PDFPlugin::debugDescription const): Deleted.
(WebKit::PDFPlugin::scrollCornerRect const): Deleted.
(WebKit::PDFPlugin::enclosingScrollableArea const): Deleted.
(WebKit::PDFPlugin::scrollableAreaBoundingBox const): Deleted.
(WebKit::PDFPlugin::isActive const): Deleted.
(WebKit::PDFPlugin::forceUpdateScrollbarsOnMainThreadForPerformanceTesting const): Deleted.
(WebKit::PDFPlugin::scrollPosition const): Deleted.
(WebKit::PDFPlugin::minimumScrollPosition const): Deleted.
(WebKit::PDFPlugin::maximumScrollPosition const): Deleted.
(WebKit::PDFPlugin::scrollbarStyleChanged): Deleted.
(WebKit::PDFPlugin::deviceScaleFactor const): Deleted.
(WebKit::PDFPlugin::calculateSizes): Deleted.
(WebKit::PDFPlugin::willDetachRenderer): Deleted.
(WebKit::PDFPlugin::updateControlTints): Deleted.
(WebKit::PDFPlugin::setScrollOffset): Deleted.
Move ScrollableArea implementation to PDFPluginBase.

(WebKit::PDFPlugin::updateScrollbars):
(WebKit::PDFPlugin::createScrollbar):
(WebKit::PDFPlugin::destroyScrollbar):
Keep scrollbar *layer* code in PDFPlugin.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::scrollPositionForTesting const):
(WebKit::PDFPluginBase::willDetachRenderer): Deleted.
(WebKit::PDFPluginBase::updateControlTints): Deleted.

(WebKit::PDFPluginBase::wantsWheelEvents const):
Turn on wantsWheelEvents, since both subclasses do.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::updateControlTints):
(WebKit::PDFPluginBase::scrollCornerRect const):
(WebKit::PDFPluginBase::enclosingScrollableArea const):
(WebKit::PDFPluginBase::scrollableAreaBoundingBox const):
(WebKit::PDFPluginBase::setScrollOffset):
(WebKit::PDFPluginBase::isActive const):
(WebKit::PDFPluginBase::forceUpdateScrollbarsOnMainThreadForPerformanceTesting const):
(WebKit::PDFPluginBase::scrollPosition const):
(WebKit::PDFPluginBase::minimumScrollPosition const):
(WebKit::PDFPluginBase::maximumScrollPosition const):
(WebKit::PDFPluginBase::deviceScaleFactor const):
(WebKit::PDFPluginBase::scrollbarStyleChanged):
(WebKit::PDFPluginBase::convertFromScrollbarToContainingView const):
(WebKit::PDFPluginBase::convertFromContainingViewToScrollbar const):
(WebKit::PDFPluginBase::debugDescription const):
(WebKit::PDFPluginBase::willDetachRenderer):
(WebKit::PDFPluginBase::updateScrollbars):
(WebKit::PDFPluginBase::createScrollbar):
(WebKit::PDFPluginBase::destroyScrollbar):
Hoist ScrollableArea implementation to here from PDFPlugin.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::geometryDidChange):
(WebKit::UnifiedPDFPlugin::updateLayout):
Fold updateLayout/updateLayerHierarchy into updateLayout(), and also update scrollbars from there.

(WebKit::UnifiedPDFPlugin::scheduleRenderingUpdate):
Added; triggers a layer flush.

(WebKit::UnifiedPDFPlugin::isLocked const):
(WebKit::UnifiedPDFPlugin::contentsSize const):
(WebKit::UnifiedPDFPlugin::firstPageHeight const):
Implement these new getters.

(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):
Move the contents layer around (for now) when the scroll offset changes.

(WebKit::UnifiedPDFPlugin::invalidateScrollbarRect):
(WebKit::UnifiedPDFPlugin::invalidateScrollCornerRect):
For now, ignor scrollbar invalidation; we don't have scrollbar layers yet.

(WebKit::UnifiedPDFPlugin::handleWheelEvent):
Pass wheel events to ScrollableArea.

Canonical link: https://commits.webkit.org/270358@main
